demos/projects/ESPRESSIF/aziotkit/main/azure_iot_freertos_esp32_main.c

```diff
@@ -0,0 +1,405 @@
+
+static void wifi_stop(void)
+{
+    esp_netif_t *wifi_netif = get_example_netif_from_desc("sta");
+    ESP_ERROR_CHECK(esp_event_handler_unregister(WIFI_EVENT,
+        WIFI_EVENT_STA_DISCONNECTED, &on_wifi_disconnect));
+    ESP_ERROR_CHECK(esp_event_handler_unregister(IP_EVENT,
+        IP_EVENT_STA_GOT_IP, &on_got_ip));
+#ifdef CONFIG_EXAMPLE_CONNECT_IPV6
+    ESP_ERROR_CHECK(esp_event_handler_unregister(IP_EVENT,
+        IP_EVENT_GOT_IP6, &on_got_ipv6));
+    ESP_ERROR_CHECK(esp_event_handler_unregister(WIFI_EVENT,
+        WIFI_EVENT_STA_CONNECTED, &on_wifi_connect));
+#endif
+    esp_err_t err = esp_wifi_stop();
+    if (err == ESP_ERR_WIFI_NOT_INIT)
+    {
+        return;
+    }
+    ESP_ERROR_CHECK(err);
+    ESP_ERROR_CHECK(esp_wifi_deinit());
+    ESP_ERROR_CHECK(esp_wifi_clear_default_wifi_driver_and_handlers(wifi_netif));
+    esp_netif_destroy(wifi_netif);
+}
+/*-----------------------------------------------------------*/
+
+static void stop(void)
+{
+    wifi_stop();
+}
+/*-----------------------------------------------------------*/
+
+static esp_err_t example_connect(void)
+{
+    if (s_semph_get_ip_addrs != NULL)
+    {
+        return ESP_ERR_INVALID_STATE;
+    }
+
+    esp_netif_t* esp_netif = wifi_start();
+    (void)esp_netif;
+
+    /* create semaphore if at least one interface is active */
+    s_semph_get_ip_addrs = xSemaphoreCreateCounting(NR_OF_IP_ADDRESSES_TO_WAIT_FOR, 0);
+
+    ESP_ERROR_CHECK(esp_register_shutdown_handler(&stop));
+    ESP_LOGI(TAG, "Waiting for IP(s)");
+    for (int i = 0; i < NR_OF_IP_ADDRESSES_TO_WAIT_FOR; ++i)
+    {
+        xSemaphoreTake(s_semph_get_ip_addrs, portMAX_DELAY);
+    }
+    // iterate over active interfaces, and print out IPs of "our" netifs
+    esp_netif_t *netif = NULL;
+    esp_netif_ip_info_t ip;
+    for (int i = 0; i < esp_netif_get_nr_of_ifs(); ++i)
+    {
+        netif = esp_netif_next(netif);
+        if (is_our_netif(TAG, netif))
+        {
+            ESP_LOGI(TAG, "Connected to %s", esp_netif_get_desc(netif));
+            ESP_ERROR_CHECK(esp_netif_get_ip_info(netif, &ip));
+
+            ESP_LOGI(TAG, "- IPv4 address: " IPSTR, IP2STR(&ip.ip));
+        }
+    }
+    return ESP_OK;
+}
+/*-----------------------------------------------------------*/
+
+static void time_sync_notification_cb(struct timeval *tv)
+{
+    ESP_LOGI(TAG, "Notification of a time synchronization event");
+    g_timeInitialized = true;
+}
+/*-----------------------------------------------------------*/
+
+static void initialize_time()
+{
+    sntp_setoperatingmode(SNTP_OPMODE_POLL);
+    sntp_setservername(0, SNTP_SERVER_FQDN);
+    sntp_set_time_sync_notification_cb(time_sync_notification_cb);
+    sntp_init();
+
+    ESP_LOGI(TAG, "Waiting for time synchronization with SNTP server");
+
+    while (!g_timeInitialized)
+    {
+        vTaskDelay(pdMS_TO_TICKS(1000));
+    }
+}
+
+/*-----------------------------------------------------------*/

```

Comment:
The next 4 functions implement the interface in `sample_azure_iot_pnp/sample_azure_iot_pnp.h`, so data can be exchanged with `sample_azure_iot_pnp/sample_azure_iot_pnp_core.c`